src/d3d9/d3d9_device.cpp

```diff
@@ -4875,44 +4895,140 @@ namespace dxvk {
+
+    Rc<DxvkBuffer>& floatBuffer = constSet.swvpBuffers.floatBuffer;
+    if (likely(constSet.meta.maxConstIndexF != 0 || floatBuffer == nullptr)) {
+      DxvkBufferSliceHandle slice;
+      if (unlikely(floatBuffer == nullptr || floatBuffer->info().size != floatBufferSize)) {
+        floatBuffer = CreateConstantBuffer(ssbo, floatBufferSize, DxsoProgramType::VertexShader, DxsoConstantBuffers::VSFloatConstantBuffer);
+        slice = floatBuffer->getSliceHandle();
+      } else {
+        slice = floatBuffer->allocSlice();
+        EmitCs([
+          cBuffer = floatBuffer,
+          cSlice  = slice
+        ] (DxvkContext* ctx) {
+          ctx->invalidateBuffer(cBuffer, cSlice);
+        });
+      }
+      auto dst = reinterpret_cast<uint8_t*>(slice.mapPtr);
+      // Max copy source size is 8192 * 64 => always aligned to any plausible value
+      // => we won't copy out of bounds
+      std::memcpy(dst, Src.fConsts, floatBufferSize);
+    }
+
+    Rc<DxvkBuffer>& intBuffer = constSet.swvpBuffers.intBuffer;
+    if (likely(constSet.meta.maxConstIndexI != 0 || intBuffer == nullptr)) {
+      DxvkBufferSliceHandle slice;
+      if (unlikely(intBuffer == nullptr || intBuffer->info().size != intBufferSize)) {
+        intBuffer = CreateConstantBuffer(ssbo, intBufferSize, DxsoProgramType::VertexShader, DxsoConstantBuffers::VSIntConstantBuffer);
+        slice = intBuffer->getSliceHandle();
+      } else {
+        slice = intBuffer->allocSlice();
+        EmitCs([
+          cBuffer = intBuffer,
+          cSlice  = slice
+        ] (DxvkContext* ctx) {
+          ctx->invalidateBuffer(cBuffer, cSlice);
+        });
+      }
+      auto dst = reinterpret_cast<uint8_t*>(slice.mapPtr);
+      // Max copy source size is 2048 * 64 => always aligned to any plausible value
+      // => we won't copy out of bounds
+      std::memcpy(dst, Src.iConsts, intBufferSize);
+    }
+
+    Rc<DxvkBuffer>& boolBuffer = constSet.swvpBuffers.boolBuffer;
+    if (likely(constSet.meta.maxConstIndexB != 0 || boolBuffer == nullptr)) {
+      DxvkBufferSliceHandle slice;
+      if (unlikely(boolBuffer == nullptr || boolBuffer->info().size != boolBufferSize)) {
+        boolBuffer = CreateConstantBuffer(ssbo, boolBufferSize, DxsoProgramType::VertexShader, DxsoConstantBuffers::VSBoolConstantBuffer);
+        slice = boolBuffer->getSliceHandle();
+      } else {
+        slice = boolBuffer->allocSlice();
+        EmitCs([
+          cBuffer = boolBuffer,
+          cSlice  = slice
+        ] (DxvkContext* ctx) {
+          ctx->invalidateBuffer(cBuffer, cSlice);
+        });
+      }
+      auto dst = reinterpret_cast<uint8_t*>(slice.mapPtr);
+      if (likely(boolBufferSize <= Layout.bitmaskSize())) {
+        std::memcpy(dst, Src.bConsts, boolBufferSize);
+      } else {
+        std::memcpy(dst, Src.bConsts, boolDataSize);
+        std::memset(dst + boolDataSize, 0, boolBufferSize - boolDataSize);
+      }
+    }
```

Comment:
This is basically the same thing copy-pasted three time, I'd create a helper function.